cranelift/isle/isle/src/serialize.rs

```diff
@@ -0,0 +1,523 @@
+                            child.set_ready(source, Readiness::Matched);
+
+                            let constraint = get_constraint(g[0]);
+                            let bindings = Vec::from_iter(
+                                constraint
+                                    .bindings_for(source)
+                                    .into_iter()
+                                    .map(|b| child.rules.find_binding(&b)),
+                            );
+
+                            let mut changed = false;
+                            for &binding in bindings.iter() {
+                                if let Some(binding) = binding {
+                                    // Matching a pattern makes its bindings available, and also emits code to bind them.
+                                    child.set_ready(binding, Readiness::Emitted);
+                                    changed = true;
+                                }
+                            }
+
+                            // As an optimization, only propagate availability if we changed any binding's readiness.
+                            if changed {
+                                child.add_bindings();
+                            }
+
+                            let body = child.sort(g);
+
+                            arms.push(MatchArm {
+                                constraint,
+                                bindings,
+                                body,
+                            });
+                        }
+
+                        Condition::Match { source, arms }
+                    }
+
+                    MatchKind::Equal(a, b) => {
+                        self.use_expr(a);
+                        self.use_expr(b);
+                        self.add_bindings();
+
+                        let mut child = self.new_block();
+                        child.set_ready(a, Readiness::Matched);
+                        child.set_ready(b, Readiness::Matched);
+                        let body = child.sort(this);
+
+                        Condition::Equal { a, b, body }
+                    }
+
+                    MatchKind::Iterator(source) => {
+                        let result = self
+                            .rules
+                            .find_binding(&Binding::Iterator { source })
+                            .unwrap();
+                        self.use_expr(source);
+                        self.add_bindings();
+
+                        let mut child = self.new_block();
+                        child.set_ready(source, Readiness::Matched);
+                        child.set_ready(result, Readiness::Emitted);
+                        child.add_bindings();
+                        let body = child.sort(this);
+
+                        Condition::Loop { result, body }
+                    }
+                };
+
+                self.finish_case(check);
+            } else {
+                debug_assert_eq!(
```

Comment:
Comment here to describe how many rules may reach this point: if not a multi-term, and if we've passed the overlap checker, then we should only have one rule if there is nothing to split on. But in a multi-term we may just be left with a bunch of rules to evaluate in priority order